hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/DirectStoreFSWriteStrategy.java

```diff
@@ -0,0 +1,258 @@
+   * for the tracking logic of PersisedStoreFileManager later on <code>commitDaughterRegion</code>
+   * method.
+   * @param parentRegionInfo {@link RegionInfo} of the parent split region.
+   * @param daughterRegionInfo {@link RegionInfo} of the resulting split region.
+   * @param familyName Column Family Name
+   * @param f File to split.
+   * @param splitRow Split Row
+   * @param top True if we are referring to the top half of the hfile.
+   * @param splitPolicy A split policy instance; be careful! May not be full populated; e.g. if
+   *                    this method is invoked on the Master side, then the RegionSplitPolicy will
+   *                    NOT have a reference to a Region.
+   * @param fs FileSystem instance for creating the actual reference file.
+   * @return
+   * @throws IOException
+   */
+  @Override
+  public Path splitStoreFile(RegionInfo parentRegionInfo, RegionInfo daughterRegionInfo,
+    String familyName, HStoreFile f, byte[] splitRow,
+    boolean top, RegionSplitPolicy splitPolicy, FileSystem fs) throws IOException {
+    Path path = super.splitStoreFile(parentRegionInfo, daughterRegionInfo,
+      familyName, f, splitRow, top, splitPolicy, fs);
+    Map<String,List<Path>> splitReferences = regionSplitReferences.
+      get(daughterRegionInfo.getEncodedName());
+    if(splitReferences==null){
+      splitReferences = new HashMap<>();
+      regionSplitReferences.put(daughterRegionInfo.getEncodedName(), splitReferences);
+    }
+    List<Path> references = splitReferences.get(familyName);
+    if(references==null){
+      references = new ArrayList<>();
+      splitReferences.put(familyName,references);
+    }
+    references.add(path);
+    return path;
+  }
+
+  /**
+   * Wraps <code>super.mergeStoreFile</code>, so that it can map the resulting merge reference to
+   * the related merged region and column family, in order to add this to 'storefile' system table
+   * for the tracking logic of PersisedStoreFileManager later in <code>commitMergedRegion</code>.
+   * @param regionToMerge {@link RegionInfo} for one of the regions being merged.
+   * @param mergedRegion {@link RegionInfo} of the merged region
+   * @param familyName Column Family Name
+   * @param f File to create reference.
+   * @param mergedDir
+   * @param fs FileSystem instance for creating the actual reference file.
+   * @return
+   * @throws IOException
+   */
+  @Override
+  public Path mergeStoreFile(RegionInfo regionToMerge, RegionInfo mergedRegion, String familyName,
+      HStoreFile f, Path mergedDir, FileSystem fs) throws IOException {
+//    Path path = super.mergeStoreFile(regionToMerge, mergedRegion, familyName, f, mergedDir, fs);
+
+    Path path = (this.fileSystem.regionInfoForFs.equals(regionToMerge)) ?
+      super.mergeStoreFile(mergedRegion, regionToMerge, familyName, f, mergedDir, fs)
+      : super.mergeStoreFile(regionToMerge, mergedRegion, familyName, f, mergedDir, fs);
```

Comment:
Couldn't we require the regions to be positional (regionA, regionB) so we didn't have to do this check and flip-flop the argument order?

Also, is n-way region merges handled at a higher level?

Reply:
> Couldn't we require the regions to be positional (regionA, regionB) so we didn't have to do this check and flip-flop the argument order?

We could assume that since this is the direct store approach, the `HRegionFileSystem` instance delegating the job for this strategy is wrapping a `resulting merged region` and the correct order of params are then passed. 

> Also, is n-way region merges handled at a higher level?

Ain't sure I follow you. You mean multiple regions merge?